src/expr/parser.rs

```diff
@@ -25,34 +26,85 @@ named!(i64_literal<i64>,
   )
 );
 
-named!(factor<i64>,
+named!(literal<Expr>,
+  map!(
+    isize_digit,
+    Expr::Literal
+  )
+);
+
+named!(factor<Expr>,
   alt!(
-    i64_literal
+    literal
   | parens
   )
 );
 
-// we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
-    mut acc: factor  ~
+    mut acc: factor ~
              many0!(
                alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
+                 map!(preceded!(tag!("*"), factor),
```

Comment:
Okay, so you switched to `map!`, which means that you're taking ownership of `add`. So far so good!

Now your central problem is how to avoid moving `acc` into the closure, because if you do this, then `acc` will not be usable in the future.